packages/gatsby/src/db/node-tracking.js

```diff
@@ -15,30 +15,71 @@ const getRootNodeId = node => rootNodeMap.get(node)
  * @see trackInlineObjectsInRootNode
  * @param {(Object|Array)} data Inline object or array
  * @param {string} nodeId Id of node that contains data passed in first parameter
+ * @param {boolean} sanitize Wether to strip objects of unuspported and not serializable fields
+ * @param {string} [ignore] Fieldname that doesn't need to be tracked and sanitized
+ *
  */
-const addRootNodeToInlineObject = (data, nodeId) => {
-  if (_.isPlainObject(data) || _.isArray(data)) {
-    _.each(data, o => addRootNodeToInlineObject(o, nodeId))
-    rootNodeMap.set(data, nodeId)
+const addRootNodeToInlineObject = (data, nodeId, sanitize, isNode = false) => {
```

Comment:
if possible I would refactor this function into 2 functions, one that sanitizes and another one that checks the nodeTracking, I think it will improve readability

Reply:
definitely possible, but this would make it so deep object traversal need to be done 2 times instead of one :S

Reply:
and most of complexity is coming here from sanitization/normalization code and not node tracking (which truly is just just:
```
    if (!isNode) {
      rootNodeMap.set(data, nodeId)
    }
```
part (other than traversal code)